dns-controller/pkg/dns/dnscache.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+)
+
+// dnsCache is a wrapper around the DNS provider, adding some caching
+type dnsCache struct {
+	// zones is the DNS provider
+	zonesProvider dnsprovider.Zones
+
+	// mutex protects the following mutable state
+	mutex sync.Mutex
+
+	cachedZones          []dnsprovider.Zone
```

Comment:
dnsprovider.Zone is an interface, so it is a reference type.  Yes, this is annoying.